### ISSUE TITLE
Training set up questions

### ISSUE THREAD

**Issue by nviolante25** [Stability-AI/stable-virtual-camera#42]
Hello,
Thank you for sharing your work!

I'm thinking of preparing a training script based on https://github.com/Stability-AI/stable-virtual-camera/issues/27 

Regarding data, I have access to DL3DV (with the COLMAP files as well). Here are some questions I have:
- Did you convert DL3DV to WebDataset format?
- Did you use a particular partition of the DL3DV 10K scenes for train/test? I see in section 4.1. you mention DL3DV140 for evaluation
- Did you directly use the COLMAP poses?
- The paper states that the cameras where normalized to [-2,2]. Are the cameras for the complete scene (all frames) normalized,and then the images sampled, or first the images are sampled and the cameras of those images are normalized?
- The paper says "We find it important to jointly sample I with a smaller sub-sampling stride to ensure sufficient temporal granularity and avoid missing critical transitions with a small probability (0.2 is used in practice)" Could you please clarify what this stride means for sampling the images?

Best regards

**Comment by jensenz-sai** [comment]
Hi @nviolante25,

Thanks for the interest in Seva! Your training script would be extremely helpful! Feel free to make a pull request.

- Yes! We use `webdataset` library with each `.tar` containing a couple of preprocessed scenes.
- We use all DL3DV `1-7K` as training split (Only these are available by the time we download, but feel free to use all splits). `DL3DV140` scenes should be exclusive from the training scenes.
- Yes!
- The normalization happens for each forward, not scene-level. Therefore, nothing fancy needs to be done during pre-processing.
- With `stride` being 1, we are sampling without subsampling frames (using consecutive frames from original clip). With `stride ` being max, we conduct random sampling. 

Hope that helps!

**Comment by nviolante25** [comment]
Hi,
Thank you for your response!

Regarding the last question, if I got it correctly 20% of the time you sample with stride=1 (consecutive frames) and the rest of the time you use random sampling?

And one extra thing I forgot to ask, did you convert the COLMAP poses in OpenCV convention to OpenGL? (this is done in the demo code for the Dust3r poses).

Best regards

**Comment by jensenz-sai** [comment]
Hi @nviolante25,

- Yes.
- Yes.
- The model takes in Plücker coordinates converted from OpenCV c2w and intrinsic. The `transforms.json` is stored with OpenGL format, and the `ReconfusionParser` will convert it back to OpenCV format for inference.
-  No.

Hope it helps!

**Comment by nviolante25** [comment]
Hi @jensenz-sai ,

Thank you for the feedback! I'll open a PR once a fix some remaining issues.

**Cameras and Dataset**
At the moment I have code to load the DL3DV data with COLMAP cameras (I have access to the COLMAP files so I can load them directly instead of using the `transforms.json` and undoing the canera conversion. (Not using WebDataset for now). 
- My question about the scaling of the cameras is what function is exactly  used. I see [these two functions](https://github.com/Stability-AI/stable-virtual-camera/blob/main/seva/eval.py#L1177) used in the demo, but I don't see how this ensures that all cameras are in $[-2,2]^3$. And also there's no explicit normalization such that the camera poses are relative to the first input frame, instead is done by centering the cameras according to their median.

**Configuration files and Conditioning**
I integrated the `sgm` library in the codebase and I've been working on getting the config file right but I'm struggling to match the `input_keys` expected in the file and the inputs the model expect according to the paper.
According to the paper, the inputs to the diffusion model are:
- concatenation of input frames, binary mask, and plucker coordinates
- concatenation of (noisy) target frames, binary mask, and plucker coordinates
- From figure 4, these two form a single batch that is the input to the model, and it outputs a cleaner version of the target frames, which is used to compute the loss.
- As extra conditioning signal we have the average CLIP embedding of the input frames.

I want to map these operations to the config files in the `GeneralConditioner` with the proper `input_key`.  I'm thinking that I need to define a new encoder to do the plucker embedding after the frames are in latent space. Currently, my dataset outputs this dict.

- My question is how to input both the target (noisy latent + plucker + mask) and input (clean latent + plucker + mask) to the network, while using the clean target frames for the loss computation. Also, did you just use the provided encoders for conditioning?
- 
At the moment, the output of my DataLoader (still need to compute Plucker embeddings, but I was thinking of doing it after the frames are map to latent space)

```
output_dict = {
  "cond_frames_without_noise": cond_frames_without_noise,  # Input frames
  "cond_frames": cond_frames,  # Input frames with optional noise
  "cond_frames_mask": cond_frames_mask,  # Boolean mask for input frames
  "c2w": c2ws,  # Camera-to-world matrices
  "K": Ks,  # Intrinsic matrices
  "camera_mask": camera_mask,  # Boolean mask for input cameras (always True?)
}
```

Thank you again you for time. 

Best regards

**Comment by jensenz-sai** [comment]
Hi @nviolante25,

Thanks again for your effort on this!

- The actual logic is now somewhat inconsistent with the writing in the paper - thanks for raising the issue, and we will rectify it in the revised version! The normalization logic is: we first center the camera (cameras will have zero-mean, [here](https://github.com/Stability-AI/stable-virtual-camera/blob/d5fe8c7b5e9b91b3977bb6cef36b9ef1b1aa42ae/seva/eval.py#L1187)); we then normalize all cameras such that the source input camera has unit length being `2` ([here](https://github.com/Stability-AI/stable-virtual-camera/blob/d5fe8c7b5e9b91b3977bb6cef36b9ef1b1aa42ae/seva/eval.py#L1201), and this does not ensures all cameras falling inside the [-2, 2] cube); Lastly, we convert all cameras into the source input camera space in `get_plucker_cordinates` functions ([here](https://github.com/Stability-AI/stable-virtual-camera/blob/d5fe8c7b5e9b91b3977bb6cef36b9ef1b1aa42ae/seva/geometry.py#L145)).
- My recommendation is to have `clean latent`, `mask`, and `plucker` from the dataloader. Map `clean latent` to `input_key` so it will be added noise in the `DiffusionEngine` and be taken as gt directly. Convert `clean latent` and `mask` into a new condition (say, it is called `replace`) such that it will operate directly on the noisy input latent and "replace" the input slots into clean latent.
- `camera_mask` is always all true.

Hope it helps.

**Comment by XinyaChen21** [comment]
I noticed in [this line](https://github.com/Stability-AI/stable-virtual-camera/blob/d5fe8c7b5e9b91b3977bb6cef36b9ef1b1aa42ae/seva/eval.py#L1178) that centering and normalization are performed using all cameras at the scene level. However, in one of your responses, you mentioned that "the normalization happens for each forward, not scene-level," which left me a bit confused. Does this mean clip-level normalization is used during training, while scene-level normalization is applied during testing?

Also, for random sampling: are the frames sampled at equal intervals? And are the sampled frames sorted in temporal order afterward?

**Comment by jensenz-sai** [comment]
Re random sampling - it is as simple as `np.random.choice(range(self.num_frames_in_seq), size=self.num_frames, replace=False)`.

**Comment by nviolante25** [comment]
Hi, @jensenz-sai  
I solved most of the things from my previous message. I added the masking and plucker conditioning and masked the input tensor in the `denoiser.py` file. However, there are some issues with the forward pass of the main SEVA network to handle a batch of frames.
Using `B F C H W` as input dim to the model produce shape errors in the `TimestepEmbedSequential` due to the usage of `Conv2D` layers, but a reshape to `(B * F) C H W` also get dimension errors [here ](https://github.com/Stability-AI/stable-virtual-camera/blob/main/seva/modules/layers.py#L126). Could you clarify what's the input format you use to handle a batch of frames? The demo doesn't include a batch dimension.

Best regards

**Comment by XinyaChen21** [comment]
Hi @jensenz-sai, may I ask whether you trained the model using fp32, fp16, or bf16?

**Comment by jensenz-sai** [comment]
Hi @nviolante25,

Since it's a spatial conv `f` should be flattened into `b`. You possibly didn't flatten the `dense_vector` to get the same size.

**Comment by jensenz-sai** [comment]
Hi @XinyaChen21,

We are usign `fp16`.

**Comment by jensenz-sai** [comment]
Hi @nviolante25,

Re-loss: we will manually assign the smallest sigma for loss weighting for the clean latent (target frames).

**Comment by nviolante25** [comment]
Hi @jensenz-sai ,

Thank you for the suggestion, I got all shapes correct. The forward pass only works with 576x576 images but images from DL3DV are not natively squared. What's you strategy to deal with that? A center crop? Resize to 576x576 and add a `original_size_as_tuple` as conditioning? I see that the `demo.py` code actually resizes the images, is this same strategy used during training?

Regarding the message from @XinyaChen21 : from what I understood from your last message I should load all the cameras of the scene (~100s for DL3DV scenes) and use all of them to recenter the cameras. Then I sample the input/target cameras and select one of them (maybe the first one?) as the source and normalize the scale. Is this correct? If that's it, I can can finish the dataloader.

Since I didn't get exactly how to use the `replace` key (the output of my dataloader are the frames, not the latents, so I can't concatenate the latents with the mask), I directly used a `mask` key from the dataloader to perform the masking in the denoiser. I think it should be equivalent. 
```
if "mask" in cond:
    mask = cond.pop("mask")[...,None,None,None].float()
    input = input * (1 - mask) + input * mask
```
 Also, was any of the CFG classes in `sampling.py` used during training for CFG? It is mentioned in the paper, and your `VanillaCFG` is different than the class from the training repo.

And finally, could you please elaborate on this?
> Re-loss: we will manually assign the smallest sigma for loss weighting for the clean latent (target frames).

Where do you get the smallest sigma? I was thinking of zeroing out the loss [here](https://github.com/Stability-AI/generative-models/blob/main/sgm/modules/diffusionmodules/loss.py#L89) for the input frames based on `cond["mask"]`

Again, thanks for your help!

**Comment by Sibo2rr** [comment]
> Hi [@jensenz-sai](https://github.com/jensenz-sai) ,
> 
> Thank you for the suggestion, I got all shapes correct. The forward pass only works with 576x576 images but images from DL3DV are not natively squared. What's you strategy to deal with that? A center crop? Resize to 576x576 and add a `original_size_as_tuple` as conditioning? I see that the `demo.py` code actually resizes the images, is this same strategy used during training?
> 
> Regarding the message from [@XinyaChen21](https://github.com/XinyaChen21) : from what I understood from your last message I should load all the cameras of the scene (~100s for DL3DV scenes) and use all of them to recenter the cameras. Then I sample the input/target cameras and select one of them (maybe the first one?) as the source and normalize the scale. Is this correct? If that's it, I can can finish the dataloader.
> 
> Since I didn't get exactly how to use the `replace` key (the output of my dataloader are the frames, not the latents, so I can't concatenate the latents with the mask), I directly used a `mask` key from the dataloader to perform the masking in the denoiser. I think it should be equivalent.
> 
> ```
> if "mask" in cond:
>     mask = cond.pop("mask")[...,None,None,None].float()
>     input = input * (1 - mask) + input * mask
> ```
> 
> Also, was any of the CFG classes in `sampling.py` used during training for CFG? It is mentioned in the paper, and your `VanillaCFG` is different than the class from the training repo.
> 
> And finally, could you please elaborate on this?
> 
> > Re-loss: we will manually assign the smallest sigma for loss weighting for the clean latent (target frames).
> 
> Where do you get the smallest sigma? I was thinking of ignoring the loss [here](https://github.com/Stability-AI/generative-models/blob/main/sgm/modules/diffusionmodules/loss.py#L89) for the input frames based on `cond["mask"]` and only compute it for the target frames. My loss configuration is:
> 
> ```
> loss_fn_config:
>   target: sgm.modules.diffusionmodules.loss.StandardDiffusionLoss
>   params:
>     loss_weighting_config:
>       target: sgm.modules.diffusionmodules.loss_weighting.EpsWeighting
>     sigma_sampler_config:
>       target: sgm.modules.diffusionmodules.sigma_sampling.DiscreteSampling
>       params:
>         num_idx: 1000
> 
>         discretization_config:
>           target: sgm.modules.diffusionmodules.discretizer.LegacyDDPMDiscretization
> ```
> 
> Again, thanks for your help!

Hi! I noticed that this part of the code uses input = input * (1 - mask) + input * mask. Just wanted to check—could there should be the mask to noise_input instead (e.g., noise_input = noise_input * (1 - mask) + input * mask)? I might be misunderstanding the logic here, so please feel free to correct me if I missed anything!

**Comment by nviolante25** [comment]
Hi @Sibo2rr ,
Thanks for pointing that out.  @jensenz-sai could you please clarify what the `replace` operation is meant to do in here? I think it's making sure that for the input frames we always use the clean latents (not the noisy version). Is that correct?

```
if "replace" in cond:
    x, mask = cond.pop("replace").split((input.shape[1], 1), dim=1)
    input = input * (1 - mask) + x * mask
```

**Comment by jensenz-sai** [comment]
@nviolante25,

Yup, you are correct - `mask` is `True` for the input frames slot. `input` is the noisy input to the network.

**Comment by nviolante25** [comment]
@jensenz-sai 

I have most of the code ready to start testing

I set the `ucg_rate=0.1` for all conditioning signals
```
- is_trainable: False
  ucg_rate: 0.1
  input_key: plucker
  target: sgm.modules.encoders.modules.IdentityEncoder

- is_trainable: False
  ucg_rate: 0.1
  input_key: concat
  target: sgm.modules.encoders.modules.IdentityEncoder

- is_trainable: False
  ucg_rate: 0.1
  input_key: replace
  target: sgm.modules.encoders.modules.IdentityEncoder

- is_trainable: False # crossattn
  ucg_rate: 0.1
  input_keys: ["frames", "mask"]
  target: sgm.modules.encoders.modules.SevaFrozenOpenCLIPImageEmbedder
  params:
    unsqueeze_dim: True
```

In summary,
- `plucker` is the `dense_vector` conditioning used [here ](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/modules/diffusionmodules/wrappers.py#L58)
- `concat` (created [here](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/data/dataset.py#L253)) is used [here](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/modules/diffusionmodules/wrappers.py#L52) in the wrapper to input the latents and the binary mask 
- `replace` (created [here](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/data/dataset.py#L266)) and used [here](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/modules/diffusionmodules/denoiser.py#L37) to replace the noisy input latent with clean latents.
- `frames` and `mask` are used to obtain the average CLIP embedding [here](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/modules/encoders/modules.py#L1057)

My (hopefully!) final question is about the guidance modules. I see in the SEVA codebase that you have defined [these](https://github.com/nviolante25/stable-virtual-camera/blob/training/seva/sampling.py#L216) custom modules. Did you use them to guide the training or simply used [these](https://github.com/nviolante25/stable-virtual-camera/blob/training/sgm/modules/diffusionmodules/guiders.py#L24) instead?

Thanks again for your assistance these last days, I really appreciate it!

**Comment by nviolante25** [comment]
Hi, thanks for the clarifications! @jensenz-sai

**Comment by nviolante25** [comment]
Hi @jensenz-sai ,
Could you please elaborate on this?

> We shift the signal-to-noise ratio (SNR) in all stages as more noise is necessary to destroy the information when training with more frames

Thanks again

**Comment by XinyaChen21** [comment]
Hi @jensenz-sai, You mentioned that the frames are re-weighted based on their distance to the nearest input. I was wondering — do you determine the nearest input and compute the distances based on the frame index in the original sequence, or based on the index within the current window? Thanks again!

**Comment by nviolante25** [comment]
Hi @yuefeng21, you need to run the SD 2.1 autoencoder on each frame of the dataset

**Comment by xiaoc57** [comment]
Hello. Although I have referred to your discussion, I am still not sure what the final weight should be. If I use sigma ** -2.0, when the timestep is small, the loss will be very large. This will cause gradient oscillation and loss oscillation. When I use 1-5 linear weighting, the gradient changes very drastically. I don't know how to deal with it. Can you provide me with a correct weight for reference?

**Comment by rzhevcherkasy** [comment]
> Hello. Although I have referred to your discussion, I am still not sure what the final weight should be. If I use sigma ** -2.0, when the timestep is small, the loss will be very large. This will cause gradient oscillation and loss oscillation. When I use 1-5 linear weighting, the gradient changes very drastically. I don't know how to deal with it. Can you provide me with a correct weight for reference?

I’m hitting a very similar issue. May I ask if you have any clue to solve this, thank you!

**Comment by xiaoc57** [comment]
@rzhevcherkasy 
Hello. I've verified this myself and others and found that oscillations are normal. The output should be verified. I can weight the training loss to sigma**-2.0. With this setting, my model converges.

**Comment by rzhevcherkasy** [comment]
@xiaoc57 
Hi, thank you for your reply. May I confirm during your training, do you use SevaWeighting or EpsWeighting? And if you use SevaWeighting, it seems sigma does not work here..?

**Comment by xiaoc57** [comment]
> [@xiaoc57](https://github.com/xiaoc57) Hi, thank you for your reply. May I confirm during your training, do you use SevaWeighting or EpsWeighting? And if you use SevaWeighting, it seems sigma does not work here..?
I used EpsWeighting. Perhaps weighting the different viewpoints might yield some gains, but simple EpsWeighting works well for fine-tuning the model. 

<img width="785" height="385" alt="Image" src="https://github.com/user-attachments/assets/11bc7837-37ae-4c25-af27-2235de5b038a" />

**Comment by rzhevcherkasy** [comment]
Hi @xiaoc57,

Thank you very much for your help. I’m currently using EpsWeighting, but I noticed that the outputs became more ambiguous after fine-tuning.

I have a few follow-up questions:

In sgm, it uses LegacyDDPMDiscretization together with its own DiscreteDenoiser, which differs from the DDPMDiscretization and DiscreteDenoiser in SEVA’s demo. Did you modify the training script in your setup?

Could you share what learning rate you are using?

You also mentioned that you made some modifications to the EulerDiscreteScheduler. Could you explain what changes you applied?

Thanks again for your support!

**Comment by xiaoc57** [comment]
> Hi [@xiaoc57](https://github.com/xiaoc57),
> 
> Thank you very much for your help. I’m currently using EpsWeighting, but I noticed that the outputs became more ambiguous after fine-tuning.
> 
> I have a few follow-up questions:
> 
> In sgm, it uses LegacyDDPMDiscretization together with its own DiscreteDenoiser, which differs from the DDPMDiscretization and DiscreteDenoiser in SEVA’s demo. Did you modify the training script in your setup?
> 
> Could you share what learning rate you are using?
> 
> You also mentioned that you made some modifications to the EulerDiscreteScheduler. Could you explain what changes you applied?
> 
> Thanks again for your support!

Hello. My noise addition process is exactly the same as the inference code. The learning rate is about 1e-5. If blurring occurs, could it be due to the data? This is very error-prone. Incidentally, I didn't use PyTorchLightning, but instead adapted the method from the diffusers library. However, this didn't cause any training errors.

**Comment by rzhevcherkasy** [comment]
Hi @xiaoc57, Thank you so much for your help. I've fixed some bugs in my training code and now it works well!
